text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,159 @@
+
+Cargo will soon gain the ability to create and chain custom registries, as described in
+https://github.com/rust-lang/cargo/pull/2361 .
+Compiler's should package the source of their implementation of the standard library as a registry, which can be distributed with the compiler.
+In practice, rustc will optionally contain the source in its sysroot.
+Rustup may be able to put it there if the default download does not contain it already.
+
+## Cargo implementation
+
+The one thing this RFC requires of the upcoming registry implementation is the ability to chain registries providing defaults and fallbacks when a registry is not manually specified.
+This is used so crates not on crates.io are instead provided by the compiler.
+[Ideally, we'd put the "interface" on crates.io to help ensure compiler-specific implementations conform, but such a mechanism is not being proposed at this time.]
+
+The injection of implicit dependencies is completely defined by the rules described in the first subsection, so this subsection will focus on the meaning of the `--resolve-compiler-specific=` flag.
+
+If `src` is included in the set passed with that flag, Cargo appends a local registry with path `${$CARGO_RUSTC --print sysroot}/src` to the back of the default chain.
+In other words, if a package is in none of the user-specified registries contain a package, Cargo will look in the registry provided by the compiler.
+
+If `bin` is included in the set passed with that flag (or inferred from the default), Cargo will build a mock registry by examining the contents of the sysroot.
+Any binary in there will be added to the mock registry, with a version deduced the best Cargo can (e.g. from the version of the compiler).
+Cargo likewise will have to be conservative with other metadata, e.g. both aborting if any feature is requested of a dep that is resolved to this mock registry, and also aborting if `default-features = false` is specified in such a dep.
+The mock registry will have dead last priority in the default chain, even behind the source registry.
+The "building" of such a package in the mock registry will consist of copying the binary into the target directory.
+
+[System packages would like to build Rust libraries 1 per system packages, and for this Cargo will need to gain some understanding of prebuilt binaries.
+It is hoped that when it does, the mock registry can be removed and the use of sysroot binaries will be less of a one-off hack.]
+
+Since Cargo will copy any binaries it needs from the sysroot when packages from the mock registry are part of the build plan, it would be nice to always prevent rustc from looking in the sysroot when compiling on Cargo's behalf.
+This would prevent users using the sysroot from forgetting to specify any non-default standard library dependencies.
+A `--no-resolve-sysroot` flag could be added for this purpose, but this is not necessary.
```

Comment:
Let's either specify it or move it to end-matter. Right now I'm leaning toward disabling sysroot when the full dag is on, but there are some interactions between all these pieces that I'm still uncertain about (need to think about it a bit still).